File: app/src/main/java/com/leitor/universal/MainActivity.java
package com.leitor.universal;

import android.app.Activity;
import android.os.Bundle;
import android.widget.*;
import android.speech.tts.TextToSpeech;
import android.content.Intent;
import android.net.Uri;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Locale;

public class MainActivity extends Activity {

    TextView textContent;
    TextToSpeech tts;

    float speed = 1.0f;
    float pitch = 1.0f;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        textContent = findViewById(R.id.textContent);
        Button open = findViewById(R.id.btnOpen);
        Button play = findViewById(R.id.btnPlay);
        Button pause = findViewById(R.id.btnPause);

        SeekBar seekSpeed = findViewById(R.id.seekSpeed);
        SeekBar seekPitch = findViewById(R.id.seekPitch);

        tts = new TextToSpeech(this, status -> {
            if (status == TextToSpeech.SUCCESS) {
                tts.setLanguage(new Locale("pt", "BR"));
            }
        });

        open.setOnClickListener(v -> {
            Intent intent = new Intent(Intent.ACTION_GET_CONTENT);
            intent.setType("text/plain");
            startActivityForResult(intent, 1);
        });

        play.setOnClickListener(v -> {
            tts.setSpeechRate(speed);
            tts.setPitch(pitch);
            tts.speak(textContent.getText().toString(),
                    TextToSpeech.QUEUE_FLUSH, null, null);
        });

        pause.setOnClickListener(v -> {
            tts.stop();
        });

        seekSpeed.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                speed = progress / 100f;
                if (speed < 0.5f) speed = 0.5f;
            }
            public void onStartTrackingTouch(SeekBar seekBar) {}
            public void onStopTrackingTouch(SeekBar seekBar) {}
        });

        seekPitch.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                pitch = progress / 100f;
                if (pitch < 0.5f) pitch = 0.5f;
            }
            public void onStartTrackingTouch(SeekBar seekBar) {}
            public void onStopTrackingTouch(SeekBar seekBar) {}
        });
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == 1 && resultCode == RESULT_OK) {
            try {
                Uri uri = data.getData();
                BufferedReader reader = new BufferedReader(
                        new InputStreamReader(getContentResolver().openInputStream(uri))
                );

                StringBuilder text = new StringBuilder();
                String line;
                while ((line = reader.readLine()) != null) {
                    text.append(line).append("\n");
                }

                textContent.setText(text.toString());

            } catch (Exception e) {
                e.printStackTrace();
            }
        }
        super.onActivityResult(requestCode, resultCode, data);
    }
}

File: app/src/main/java/com/leitoruniversalia/MainActivity.kt
package com.leitoruniversalia

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.speech.tts.TextToSpeech
import androidx.appcompat.app.AppCompatActivity
import android.widget.*
import java.io.*
import java.util.*
import com.tom_roush.pdfbox.pdmodel.PDDocument
import com.tom_roush.pdfbox.text.PDFTextStripper
import org.apache.tika.Tika
import nl.siegmann.epublib.epub.EpubReader

class MainActivity : AppCompatActivity(), TextToSpeech.OnInitListener {

    private lateinit var tts: TextToSpeech
    private lateinit var editText: EditText

    private var pausedText:String=""
    private val PICK_FILE=100

    override fun onCreate(savedInstanceState:Bundle?) {

        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        editText=findViewById(R.id.editText)

        val open=findViewById<Button>(R.id.buttonOpen)
        val read=findViewById<Button>(R.id.buttonRead)
        val pause=findViewById<Button>(R.id.buttonPause)
        val resume=findViewById<Button>(R.id.buttonResume)
        val stop=findViewById<Button>(R.id.buttonStop)
        val clear=findViewById<Button>(R.id.buttonClear)
        val clearHistory=findViewById<Button>(R.id.buttonClearHistory)
        val spinner=findViewById<Spinner>(R.id.speedControl)

        tts=TextToSpeech(this,this)

        val speeds=arrayOf("0.5","0.75","1.0","1.25","1.5","2.0")
        spinner.adapter=ArrayAdapter(this,android.R.layout.simple_spinner_dropdown_item,speeds)

        spinner.onItemSelectedListener=object:AdapterView.OnItemSelectedListener{
            override fun onItemSelected(parent:AdapterView<*>?,view:android.view.View?,pos:Int,id:Long){
                val rate=speeds[pos].toFloat()
                tts.setSpeechRate(rate)
            }
            override fun onNothingSelected(parent:AdapterView<*>?){}
        }

        open.setOnClickListener{

            val intent=Intent(Intent.ACTION_GET_CONTENT)
            intent.type="*/*"
            startActivityForResult(intent,PICK_FILE)

        }

        read.setOnClickListener{

            val text=editText.text.toString()
            pausedText=text
            tts.speak(text,TextToSpeech.QUEUE_FLUSH,null,null)

            saveHistory(text)

        }

        pause.setOnClickListener{

            pausedText=editText.text.toString()
            tts.stop()

        }

        resume.setOnClickListener{

            tts.speak(pausedText,TextToSpeech.QUEUE_FLUSH,null,null)

        }

        stop.setOnClickListener{

            tts.stop()

        }

        clear.setOnClickListener{

            editText.setText("")

        }

        clearHistory.setOnClickListener{

            getSharedPreferences("history",MODE_PRIVATE).edit().clear().apply()

        }

    }

    override fun onActivityResult(requestCode:Int,resultCode:Int,data:Intent?){

        super.onActivityResult(requestCode,resultCode,data)

        if(requestCode==PICK_FILE && resultCode==Activity.RESULT_OK){

            try{

                val uri=data!!.data!!
                val input=contentResolver.openInputStream(uri)

                val name=uri.toString().lowercase()

                var text=""

                if(name.endsWith(".pdf")){

                    val doc=PDDocument.load(input)
                    val stripper=PDFTextStripper()
                    text=stripper.getText(doc)
                    doc.close()

                }else if(name.endsWith(".epub")){

                    val book=EpubReader().readEpub(input)
                    val builder=StringBuilder()

                    book.contents.forEach{

                        builder.append(String(it.data))

                    }

                    text=builder.toString()

                }else{

                    val tika=Tika()
                    text=tika.parseToString(input)

                }

                editText.setText(text)

            }catch(e:Exception){

                Toast.makeText(this,"Erro ao abrir arquivo",Toast.LENGTH_LONG).show()

            }

        }

    }

    private fun saveHistory(text:String){

        val prefs=getSharedPreferences("history",MODE_PRIVATE)
        prefs.edit().putString("last",text).apply()

    }

    override fun onInit(status:Int){

        if(status==TextToSpeech.SUCCESS){

            tts.language=Locale("pt","BR")

        }

    }

    override fun onDestroy(){

        if(::tts.isInitialized){

            tts.stop()
            tts.shutdown()

        }

        super.onDestroy()

    }

}
